flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java

```diff
@@ -137,12 +150,26 @@ public StreamStateHandle registerReference(
         return entry.stateHandle;
     }
 
+    @Override
+    public void registerCompletedCheckpoint(CompletedCheckpoint checkpoint) {
+        synchronized (registeredCheckpoints) {
+            LOG.trace("Register checkpoint {}.", checkpoint.getCheckpointID());
+            registeredCheckpoints.add(checkpoint);
+        }
+    }
+
+    @Override
+    public void setPostCleanAction(Runnable postCleanAction) {
+        this.postCleanAction = postCleanAction;
```

Comment:
I'd prefer this action to be passed as an argument to `unregisterUnusedState` rather than having a mutable field, because
1. on shutdown, it should not be called, right? with an argument, it's easier to pass a no-op callback
2. it's easy to forget to call this setter and more difficult to understand **when** is it called

Reply:
Nice suggestion. I pass the `subsumAction` as an argument to unregisterUnusedState, and subsume action contains `postCleanAction`.